Core/src/org/sleuthkit/autopsy/centralrepository/datamodel/CorrelationAttributeNormalizer.java

```diff
@@ -0,0 +1,164 @@
+                return CorrelationAttributeNormalizer.normalize(type, data);
+            } else {
+                throw new CorrelationAttributeNormalizationException(String.format("Given attributeTypeId did not correspond to any known Attribute: %s", attributeTypeId));
+            }
+        } catch (EamDbException ex) {
+            throw new CorrelationAttributeNormalizationException(ex);
+        }
+    }
+
+    /**
+     * Verify MD5 is the correct length and values. Make lower case.
+     */
+    private static String normalizeMd5(String data) throws CorrelationAttributeNormalizationException {
+        final String errorMessage = "Data purporting to be an MD5 was found not to comform to expected format: %s";
+        if(data == null){
+            throw new CorrelationAttributeNormalizationException(String.format(errorMessage, data));
+        }
+        final String validMd5Regex = "^[a-fA-F0-9]{32}$";
+        final String dataLowered = data.toLowerCase();
+        if(dataLowered.matches(validMd5Regex)){
+            return dataLowered;
+        } else {
+            throw new CorrelationAttributeNormalizationException(String.format(errorMessage, data));
+        }
+    }
+
+    /**
+     * Verify there are no slashes or invalid domain name characters (such as '?' or \: ). Normalize to lower case.
+     */
+    private static String normalizeDomain(String data) throws CorrelationAttributeNormalizationException {
+        DomainValidator validator = DomainValidator.getInstance(true);
+        if(validator.isValid(data)){
+            return data.toLowerCase();
+        } else {
+            throw new CorrelationAttributeNormalizationException(String.format("Data was expected to be a valid domain: %s", data));
+        }
+    }
+
+    /**
+     *  Verify that there is an '@' and no invalid characters. Should normalize to lower case.
+     */
+    private static String normalizeEmail(String data) throws CorrelationAttributeNormalizationException {
+        EmailValidator validator = EmailValidator.getInstance(true, true);
+        if(validator.isValid(data)){
+            return data.toLowerCase();
+        } else {
+            throw new CorrelationAttributeNormalizationException(String.format("Data was expected to be a valid email address: %s", data));
+        }
+    }
+
+    /**
+     * Verify it is only numbers and '+'. Strip spaces, dashes, and parentheses.
+     */
+    private static String normalizePhone(String data) throws CorrelationAttributeNormalizationException {
+        String phoneNumber = data.replaceAll("[^0-9\\+]", "");
+        if(phoneNumber.matches("\\+?[0-9]+")){
+            return phoneNumber;
+        } else {
+            throw new CorrelationAttributeNormalizationException(String.format("Data was expected to be a valid phone number: %s", data));
+        }
+    }
+
+    /**
+     * USB ID is of the form: hhhh:hhhh where h is a hex digit.  Convert to lower case.
```

Comment:
Is this right? I'm not sure it matches my sample data

Reply:
I found (on the dubious internet) USB IDs in a variety of formats, thus the flexible validation you see in the regex.  Can I see your test data, or do you have any suggestions on how to create my own?

Reply:
I found this site which matches more of what I see:
https://docs.microsoft.com/en-us/windows-hardware/drivers/install/standard-usb-identifiers
Mine look more like this: 4&24d6eb65&0
If we can't find a good source it might be better to skip validation than to possibly lose data. There aren't usually that many device ID attributes anyway.

Reply:
From MS: USB\VID_v(4)&PID_d(4)&REV_r(4)_

So in your interpretation of this spec, are they just dropping the insignificant digits, so it really should specify up to 4 digits and not exactly four digits?

It goes on to discuss these various INF models.  Should we be expecting and accepting those as well?